### PR TITLE
44 interface that doesnt affect rprofile

### PR DESCRIPTION
**Pull request by e-kotov**
Implementing #44 and partially #45 .

@hadley, so far without checking for existing JAVA_HOME, just pure proof of concept.

To test:

```r
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}

remotes::install_github(
  "e-kotov/rJavaEnv@44-interface-that-doesnt-affect-rprofile",
  force = TRUE, dependencies = TRUE
)

library(rJavaEnv)
getOption("rJavaEnv.valid_major_java_versions")

use_java(8)
"8" == java_check_version_cmd(quiet = TRUE)
"8" == java_check_version_rjava(quiet = TRUE)

use_java(11)
"11" == java_check_version_cmd(quiet = TRUE)
"11" == java_check_version_rjava(quiet = TRUE)

use_java(17)
"17" == java_check_version_cmd(quiet = TRUE)
"17" == java_check_version_rjava(quiet = TRUE)

use_java(21)
"21" == java_check_version_cmd(quiet = TRUE)
"21" == java_check_version_rjava(quiet = TRUE)

use_java(22)
"22" == java_check_version_cmd(quiet = TRUE)
"22" == java_check_version_rjava(quiet = TRUE)

# and back to 8, to see if cached installation setting is instant and skips download and install/unpacking
use_java(8)
"8" == java_check_version_cmd(quiet = TRUE)
"8" == java_check_version_rjava(quiet = TRUE)


# to reset downloaded files
java_clear("distrib", delete_all = TRUE)
java_clear("installed", delete_all = TRUE)

```

### CODE REVIEW

File: R/java_env.R [hunk 1]
```diff
@@ -248,13 +255,13 @@ java_check_version_cmd <- function(
   # Get JAVA_HOME again and check if it's set
   current_java_home <- Sys.getenv("JAVA_HOME")
   if (current_java_home == "") {
-    cli::cli_alert_warning("JAVA_HOME is not set.")
+    if (!quiet) cli::cli_alert_warning("JAVA_HOME is not set.")
```
Comment (hadley) [inline R/java_env.R:258]: FWIW we've mostly switched to this style: `cli::cli_inform(c("!" = "JAVA_HOME is not set."))`

File: R/java_download.R [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #' Download a Java distribution
 #'
-#' @param version The Java version to download. If not specified, defaults to the latest LTS version.
-#' @param distribution The Java distribution to download. If not specified, defaults to "Corretto".
+#' @param version `Integer` or `character` vector of length 1 for major version of Java to download or install. If not specified, defaults to the latest LTS version. Can be "8", "11", "17", "21", "22", or 8, 11, 17, 21, or 22.
```
Comment (hadley) [inline R/java_download.R:3]: IMO it's better to be stricter here and only accept numbers, not strings.

File: R/java_download.R [hunk 3]
```diff
@@ -40,18 +40,19 @@ java_download <- function(
...
   }
 
-  rje_consent_check()
+  # rje_consent_check() # disabling consent check for now
   java_urls <- java_urls_load()
 
   valid_distributions <- names(java_urls)
   valid_platforms <- names(java_urls[[distribution]])
   valid_architectures <- names(java_urls[[distribution]][[platform]])
+  valid_versions <-
```
Comment (hadley) [inline R/java_download.R:49]: Is this supposed to be on its own line?

File: R/onLoad.R [hunk 4]
```diff
@@ -3,7 +3,8 @@
   op.rJavaEnv <- list(
     # this default folder choice is in-line with what renv package does
     # https://github.com/rstudio/renv/blob/d6bced36afa0ad56719ca78be6773e9b4bbb078f/R/bootstrap.R#L940-L950
-    rJavaEnv.cache_path = tools::R_user_dir("rJavaEnv", which = "cache")
+    rJavaEnv.cache_path = tools::R_user_dir("rJavaEnv", which = "cache"),
+    rJavaEnv.valid_major_java_versions = c("8", "11", "17", "21", "22")
```
Comment (hadley) [inline R/onLoad.R:7]: It feels a bit surprising to me to put this in an option? Under what circumstances would the user want to change it?

File: R/java_env.R [hunk 5]
```diff
@@ -267,25 +274,27 @@ java_check_version_cmd <- function(
...
+#' @return A `character` vector of length 1 containing the major Java version.
 #' @keywords internal
-#'
-#' @return TRUE if successful, otherwise FALSE.
-java_check_version_system <- function() {
+#' 
+java_check_version_system <- function(
+  quiet
+) {
   which_java <- tryCatch(
```
Comment (hadley) [inline R/java_env.R:298]: Is there a reason you don't use `Sys.which()` here?

File: R/java_unpack.R [hunk 6]
```diff
@@ -0,0 +1,79 @@
+#' Unpack a Java distribution file into cache directory
+#' 
+#' @inheritParams java_install
+#' @inheritParams global_quiet_param
+#' @return A `character` vector containing of length 1 containing the path to the unpacked Java directory.
+#' @keywords internal
+java_unpack <- function(
```
Comment (hadley) [inline R/java_unpack.R:7]: Pulling this out into a separate function makes a lot of sense to me 😄

File: R/java_unpack.R [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  java_versions <- getOption("rJavaEnv.valid_major_java_versions")
+
+  # Extract information from the file name
+  filename <- basename(java_distrib_path)
+  parts <- strsplit(gsub("\\.tar\\.gz|\\.zip", "", filename), "-")[[1]]
+
+  # Guess the version, architecture, and platform
+  version <- parts[vapply(parts, function(x) x %in% java_versions, logical(1))][1]
+  arch <- parts[vapply(parts, function(x) x %in% architectures, logical(1))][1]
+  platform <- parts[vapply(parts, function(x) x %in% platforms, logical(1))][1]
```
Comment (hadley) [inline R/java_unpack.R:22]: If I'm reading the code correctly, `parts` will be a character vector so you can use `%in%` directly.

```suggestion
    version <- parts[parts %in% java_versions][1]
    arch <- parts[parts %in% architectures][1]
    platform <- parts[parts %in% platforms][1]
```

File: R/java_unpack.R [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+  if (is.na(version)) stop(cli::cli_abort("Unable to detect Java version from filename.", .envir = environment()))
+  if (is.na(arch)) stop(cli::cli_abort("Unable to detect architecture from filename.", .envir = environment()))
+  if (is.na(platform)) stop(cli::cli_abort("Unable to detect platform from filename.", .envir = environment()))
```
Comment (hadley) [inline R/java_unpack.R:26]:
```suggestion
  if (is.na(version)) cli::cli_abort("Unable to detect Java version from filename.")
  if (is.na(arch)) cli::cli_abort("Unable to detect architecture from filename.")
  if (is.na(platform)) cli::cli_abort("Unable to detect platform from filename.")
```

You only need to pass `envir` when you're using a helper that throws and error and want to make sure the error message references the function that the user actually called.

File: R/use_java.R [hunk 9]
```diff
@@ -0,0 +1,61 @@
+#' Install specified Java version and set the `JAVA_HOME` and `PATH` environment variables in current R session
+#' 
+#' @description
+#' Using specified Java version, set the `JAVA_HOME` and `PATH` environment variables in the current R session. If Java distribtuion has not been downloaded yet, download it. If it was not installed into cache directory yet, install it there and then set the environment variables. This is intended as a quick and easy way to use different Java versions in R scripts that are in the same project, but require different Java versions. For example, one could use this in scripts that are called by `targets`` package or `callr` package.
```
Comment (hadley) [inline R/use_java.R:4]:
```suggestion
#' Using specified Java version, set the `JAVA_HOME` and `PATH` environment variables in the current R session. If Java distribtuion has not been downloaded yet, download it. If it was not installed into cache directory yet, install it there and then set the environment variables. This is intended as a quick and easy way to use different Java versions in R scripts that are in the same project, but require different Java versions. For example, one could use this in scripts that are called by `targets` package or `callr` package.
```

File: R/use_java.R [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+  java_env_set(
+    where = "session",
+    java_home = java_cached_install_path,
+    quiet = quiet
+  )
+
+  if (!quiet) {
+    cli::cli_alert_success("Java version {version} was set in the current R session")
+  }
```
Comment (hadley) [inline R/use_java.R:60]:
```suggestion
  }
  invisible()
```

I think it's a good idea to return `invisible()` from side-effecty functions to indicate that there's no expected value; just some side effects.

File: R/use_java.R [hunk 11]
```diff
@@ -0,0 +1,61 @@
+#' Install specified Java version and set the `JAVA_HOME` and `PATH` environment variables in current R session
```
Comment (hadley) [inline R/use_java.R:1]: This was just what I was imagining. Thanks!

File: R/java_download.R (hunk 2 above)
Comment (e-kotov) [reply R/java_download.R:3]: I thought of supporting specific java versions beyond major version at some point. Not sure it is worth it. But if it is, strings will be necessary.

File: R/java_download.R (hunk 3 above)
Comment (e-kotov) [reply R/java_download.R]: nope. I wonder how r cmd check passed this. removed.

File: R/onLoad.R (hunk 4 above)
Comment (e-kotov) [reply R/onLoad.R:7]: Well, they just might. Also, this allows me to set the cache folder to temp directory in examples, as I was worried about CRAN review and "nowhere except temp dir" policy.

File: R/java_env.R (hunk 5 above)
Comment (e-kotov) [reply R/java_env.R:298]: I honestly did not know that `Sys.which()` existed (just like many other good base R functions). Will happily fix.

File: R/java_unpack.R (hunk 6 above)
Comment (e-kotov) [reply R/java_unpack.R:7]: To me it became obvious once I started moving things around for `use_java()`

**Comment by e-kotov** [comment]
> This was just what I was imagining. Thanks!

@hadley thanks for the review, tips and suggestions! Implementing this actually caused some code cleanup, which is also great. And the feature itself is super useful, thank for a great request! It will also work fit well with my planned vignette on how rJavaEnv can be used with `{targets}` and `{callr}`. I'll revise once more this week and submit this version to CRAN.

**Pull request merged**